### DISCUSSION TITLE
HikoGUI is now a C++23 library.

### DISCUSSION THREAD

**Discussion by takev** [hikoworks/hikogui#458] [News]
Due to the impossibility to use C++20 modules without having `import std` (can't both `#include` and `import` the same types, including types from the `std` library), HikoGUI is now a C++23 library.

I do not want to delay the use of HikoGUI, therefor I will only allow C++23 constructions that are supported by all the major compilers: MSVC, clang and gcc.

As of this moment I will allow the following C++23 constructions:

- import std
- UTF-8 as source code character set
- std::is_scoped_enum
- std::basic_string::contains
- std::to_underlying
- std::visit on classes derived from std::variant
- constexpr type_info::operator==
- Iterator pair constructors for std::stack and std::queue
- Range constructor for std::basic_string_view
- std::invoke_r
- constexpr std::optional, std::variant
- std::lazy_zip
- std::byteswap
- std::basic_string::resize_and_overwrite
- std::expected
- std::unreachable
- constexpr std::unique_ptr
- constexpr std::to_chars, std::from_chars
- views::as_rvalue
- std::basic_string_format
- std::views::zip

Any constructions that are part of HikoGUI which are also available in C++23 (`_uz` and `hi::generator`) will eventually be replaced by C++23 constructions once they are available in all the major compilers.

**Comment by jakoch** [comment]
I would suggest to keep re-implementations of `C++NN` features in their individual namespaces: `hi20::`, `hi23::`.
Basically having a language feature or compatibility namespace.
This allows to pick the "lowest NN required" and switch to `std::` implementation, when available. 
When you raise the NN level of the library, you can simply replace with `s/hi20::/std::/g` on the source and delete the implementation.
